canary/producer-cpp/src/Canary.cpp

```diff
@@ -0,0 +1,899 @@
+
+    GstCaps *caps;
+
+    // define the elements
+    source = gst_element_factory_make("videotestsrc", "source");
+    autovidcon = gst_element_factory_make("autovideoconvert", "vidconv");
+    h264enc = gst_element_factory_make("x264enc", "h264enc");
+    h264parse = gst_element_factory_make("h264parse", "h264parse");
+    appsink = gst_element_factory_make("appsink", "appsink");
+
+    // to change output video pattern to a moving ball, uncomment below
+    //g_object_set(source, "pattern", 18, NULL);
+
+    // videotestsrc must be set to "live" in order for pts and dts to be incremented
+    g_object_set(source, "is-live", TRUE, NULL);
+
+    // configure appsink
+    g_object_set(G_OBJECT (appsink), "emit-signals", TRUE, "sync", FALSE, NULL);
+    g_signal_connect(appsink, "new-sample", G_CALLBACK(on_new_sample), data);
+    
+    // define the elements
+    // h264enc = gst_element_factory_make("vtenc_h264_hw", "h264enc");
+    h264parse = gst_element_factory_make("h264parse", "h264parse");
+
+    // define and configure video filter, we only want the specified format to pass to the sink
+    // ("caps" is short for "capabilities")
+    string video_caps_string = "video/x-h264, stream-format=(string) avc, alignment=(string) au";
+    video_filter = gst_element_factory_make("capsfilter", "video_filter");
+    caps = gst_caps_from_string(video_caps_string.c_str());
+    g_object_set(G_OBJECT (video_filter), "caps", caps, NULL);
+    gst_caps_unref(caps);
+
+    video_caps_string = "video/x-raw, framerate=" + to_string(data->pCanaryConfig->testVideoFps) + "/1" + ", width=1440, height=1080";
+    video_src_filter = gst_element_factory_make("capsfilter", "video_source_filter");
+    caps = gst_caps_from_string(video_caps_string.c_str());
+    g_object_set(G_OBJECT (video_src_filter), "caps", caps, NULL);
+    gst_caps_unref(caps);
+    
+    // check if all elements were created
+    if (!pipeline || !source || !video_src_filter || !appsink || !autovidcon || !h264parse || 
+        !video_filter || !h264enc)
+    {
+        g_printerr("Not all elements could be created.\n");
+        return 1;
+    }
+
+    // build the pipeline
+    gst_bin_add_many(GST_BIN (pipeline), source, video_src_filter, autovidcon, h264enc,
+                    h264parse, video_filter, appsink, NULL);
+
+    // check if all elements were linked
+    if (!gst_element_link_many(source, video_src_filter, autovidcon, h264enc, 
+        h264parse, video_filter, appsink, NULL)) 
+    {
+        g_printerr("Elements could not be linked.\n");
+        gst_object_unref(pipeline);
+        return 1;
+    }
+
+    return 0;
+}
+
+int gstreamer_init(int argc, char* argv[], CustomData *data) {
+
+    // init GStreamer
+    gst_init(&argc, &argv);
+
+    GstElement *pipeline;
+    int ret;
+    GstStateChangeReturn gst_ret;
+
+    // Reset first frame pts
+    data->first_pts = GST_CLOCK_TIME_NONE;
+
+    switch (data->streamSource) {
+        case CustomData::TEST_SOURCE:
```

Comment:
Would suggest changing the place for definition of TEST_SOURCE outside CustomData to avoid this.